### PR TITLE
Add channel handlers to debug inbound and outbound events

### PR DESCRIPTION
**Pull request by JovanMilenkovic**
Motivation:

Users may want to log all of inbound and/or outbound events. This addresses https://github.com/apple/swift-nio/issues/475

Modifications:

Add DebugInboundEventsHandler and DebugOutboundEventsHandler

Result:

Users can plug additional handlers into their pipeline for default printing of events or getting a hook with relevant information for their own logging mechanism.

### CODE REVIEW

File: Sources/NIOExtras/DebugInboundEventsHandler.swift [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import NIO
```
Comment (Lukasa) [inline Sources/NIOExtras/DebugInboundEventsHandler.swift]: Do you mind adding our license header to the top of this file?

File: Sources/NIOExtras/DebugInboundEventsHandler.swift [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+public class DebugInboundEventsHandler: ChannelInboundHandler {
```
Comment (Lukasa) [inline Sources/NIOExtras/DebugInboundEventsHandler.swift]: Can you add a doc comment for this please?

File: Sources/NIOExtras/DebugOutboundEventsHandler.swift [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import NIO
```
Comment (Lukasa) [inline Sources/NIOExtras/DebugOutboundEventsHandler.swift]: Can you add our license header?

File: Sources/NIOExtras/DebugOutboundEventsHandler.swift [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+public class DebugOutboundEventsHandler: ChannelOutboundHandler {
```
Comment (Lukasa) [inline Sources/NIOExtras/DebugOutboundEventsHandler.swift]: This class needs a doc comment.

File: Sources/NIOExtras/DebugOutboundEventsHandler.swift [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    
+    public typealias OutboundIn = ByteBuffer
```
Comment (Lukasa) [inline Sources/NIOExtras/DebugOutboundEventsHandler.swift]: This is incorrect.

The correct two typealiases are `OutboundIn = Any` and `OutboundOut = Any`.

File: Sources/NIOExtras/DebugInboundEventsHandler.swift [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    
+    public typealias InboundIn = ByteBuffer
```
Comment (Lukasa) [inline Sources/NIOExtras/DebugInboundEventsHandler.swift]: This is incorrect. The correct two typealiases are `InboundIn = Any` and `InboundOut = Any`.

File: Sources/NIOExtras/DebugOutboundEventsHandler.swift [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        let message: String
+        switch event {
+        case .register:
+            message = "Registering channel"
+        case .bind(let address):
+            message = "Binding to \(address)"
+        case .connect(let address):
+            message = "Connecting to \(address)"
+        case .write(let data):
+            message = "Writting \(data)"
```
Comment (Lukasa) [inline Sources/NIOExtras/DebugOutboundEventsHandler.swift]: Nit: "Writing"

File: Tests/NIOExtrasTests/DebugInboundEventsHandlerTest.swift [hunk 8]
```diff
@@ -0,0 +1,111 @@
+import XCTest
+import NIO
+import NIOExtras
+
+class DebugInboundEventsHandlerTest: XCTestCase {
+    
+    private var channel: EmbeddedChannel!
+    private var lastEvent: DebugInboundEventsHandler.Event!
+    private var sut: DebugInboundEventsHandler!
```
Comment (Lukasa) [inline Tests/NIOExtrasTests/DebugInboundEventsHandlerTest.swift]: What is `sut`? This is a fairly unclear variable name.

File: Tests/NIOExtrasTests/DebugInboundEventsHandlerTest.swift [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+        case (.read(let data1), .read(let data2)):
+            return "\(data1)" == "\(data2)"
+        case (.readComplete, .readComplete):
+            return true
+        case (.writabilityChanged(let isWritable1), .writabilityChanged(let isWritable2)):
+            return isWritable1 == isWritable2
+        case (.userInboundEventTriggered(let event1), .userInboundEventTriggered(let event2)):
+            return event1 as! String == event2 as! String
+        case (.errorCaught(let error1), .errorCaught(let error2)):
+            print(error1.localizedDescription)
```
Comment (Lukasa) [inline Tests/NIOExtrasTests/DebugInboundEventsHandlerTest.swift]: Remove this print.

File: Tests/NIOExtrasTests/DebugOutboundEventsHandlerTest.swift [hunk 10]
```diff
@@ -0,0 +1,98 @@
+import XCTest
+import NIO
+import NIOExtras
+
+class DebugOutboundEventsHandlerTest: XCTestCase {
+
+    private var channel: EmbeddedChannel!
+    private var lastEvent: DebugOutboundEventsHandler.Event!
+    private var sut: DebugOutboundEventsHandler!
```
Comment (Lukasa) [inline Tests/NIOExtrasTests/DebugOutboundEventsHandlerTest.swift]: `sut` is unclear here too.

**Comment by JovanMilenkovic** [comment]
@Lukasa tnx for the review. Regarding "sut", it's supposed to be SYSTEM UNDER TEST and provide a clear distinction between the main variable and helper variables in the test suite. It's fairly common (https://github.com/search?q=sut&type=Code) and probably originates from https://www.istqb.org/downloads/send/20-istqb-glossary/210-istqb-glossary-3-2-ga-release-notes-final.html. I see now that it currently doesn't belong in NIO test suites (it was my force of habit) but perhaps it should be further explored, I think it pulls it's weight.

File: Sources/NIOExtras/DebugInboundEventsHandler.swift (hunk 2 above)
Comment (Lukasa) [reply Sources/NIOExtras/DebugInboundEventsHandler.swift]: Sorry, but doc comments begin each line with the four character string: `/// ` (that is, three slashes and then a space).

File: Sources/NIOExtras/DebugOutboundEventsHandler.swift (hunk 4 above)
Comment (Lukasa) [reply Sources/NIOExtras/DebugOutboundEventsHandler.swift]: Sorry, but doc comments begin each line with the four character string: `/// ` (that is, three slashes and then a space).

Comment (JovanMilenkovic) [reply Sources/NIOExtras/DebugOutboundEventsHandler.swift]: Ah right, sorry for not paying too much attention. Updated with real docs

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Comment by weissi** [comment]
@swift-nio-bot test this please

**Pull request merged**